### ISSUE TITLE
decorator ordering matters

### ISSUE THREAD

**Issue by blalor** [fabric/fabric#1406]
Given the following fabfile:
```python
from fabric.api import task, runs_once


@task
@runs_once
def do_thing(thing="some"):
    print "%sthing done" % thing


@runs_once
@task
def do_thing_else(thing="other"):
    print "%sthing else done" % thing
```

`fab -d do_thing` doesn't show the arguments:
```
Displaying detailed information for task 'do_thing':

    No docstring provided
    Arguments:
```

`fab -d do_thing_else` *does*, however:
```
Displaying detailed information for task 'do_thing_else':

    No docstring provided
    Arguments: thing='other'
```

Given that the ordering of the decorators doesn't impact how the task is executed, the ordering shouldn't break the details.

**Comment by blalor** [comment]
Additionally, putting `@parallel` before `@task` causes the function to not be seen as a task.

**Comment by kaaveland** [comment]
Looking into this a bit, mostly a note to self (just posting in case it helps anyone else):

TL;DR: The culmination of this is discovering that at `inner_decorator` in decorators.py:54 does not preserve the `.wrapped` attribute that `WrappedCallableTask` works with, which makes `__details__` pass `inner_decorator` instead into `get_task_details`, whereas it should have been the closed-over `func` variable from `attach_list` in decorators.py. I'm not sure what a safe fix would be, the machinery in `tasks.py` needs the `.wrapped` attribute to function correctly. Going by the comment in `_list_annotating_decorator`, `_wrap_as_new` seems to exhibit the wrong behaviour (missing a `not` in the condition)? 

```python
        # Don't replace @task new-style task objects with inner_decorator by
        # itself -- wrap in a new Task object first.
def _wrap_as_new(original, new):
    if isinstance(original, tasks.Task):
        return tasks.WrappedCallableTask(new)
    return new

```

The long story down here:

So, initially I thought this should be a simple thing, so I looked a bit at the code and played around with the interpreter. After looking at how `@task` was defined, I was pretty confident that it would always create a new `WrappedCallableTask` on top of whatever it decorated. I proved myself false right away:

```python
>>> from fabric.api import task, roles
>>> @task
... @roles('role')
... def some_task(): pass
...
# My theory is now that we should be able to follow `.wrapped` twice, but that's wrong:
>>> some_task.wrapped.wrapped
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'function' object has no attribute 'wrapped'
# Changing the order:
>>> @roles('role')
... @task
... def another_task(): pass
...
>>> another_task.wrapped.wrapped
<function another_task at 0x10d53a320>
```

So clearly, there's a difference there. My next hunch was that maybe something that was attempting to use `.wrapped` wasn't following the attribute down to the "root".

And then I looked into `WrappedCallableTask` and its `__getattr__` and at this point I must admit to being a bit stumped by my Python knowledge. Initially I thought that this might recurse:
```python
    def __getattr__(self, k):
        return getattr(self.wrapped, k)
```
But clearly that isn't true, because then it would be impossible for use to follow `.wrapped` twice. Looking at the doc I learn that this method is only used when the attribute isn't found in one of the "usual places", so it's basically `method_missing` from Ruby. So basically, anything that isn't found in our own `.wrapped` should be delegated to the next `.wrapped` and then that should do the same. So, doing the following:
```python
>>> def foo(): pass
...
>>> foo.attr = 1
>>> inner = WrappedCallableTask(foo)
>>> inner.attr2 = 2
>>> outer = WrappedCallableTask(inner)
>>> outer.attr3 = 3
```
All the 3 attr fields should be accessible on `outer` (which they are). I did notice that only `attr3` will show up in `dir(outer)`, so I figure maybe we're doing some introspection-weirdness somewhere (trying to use `__dict__` maybe?). But I can't see any usage of `__dict__` or `dir()`, so clearly that's not it.

Thinking here that I'm focusing on the wrong problem. Backtracking to decorated functions above, I can get `.roles` on both `some_task`  and `another_task`. So the problem isn't that we can't obtain the attributes and it was a waste of time to dig into `__getattr__` above (although I did learn something new, so that's nice).

So at this point it looks like it shouldn't matter that one of the ways nest `WrappedCallableTask` wrt obtaining attribute values. So maybe it's because of the way we load things. Next up, `__details__` looks like something we use internally in order to determine what a task looks like. Here we go down to the 'root' `.wrapped` to fetch details (this will always be the innermost decoratee as far as I can tell). `get_task_details` is then invoked on that, and it uses `inspect.getargs` to figure out our arguments. In the `@task` then `@roles` case, that clearly doesn't return the correct value. So yeah, turns out that we don't have the correct innermost `wrapped` here somehow, but we still have a handle to the original function from somewhere since I can invoke it. So turning to `inspect.getsourcelines` to discover what our `.wrapped` actually is, it turns out that it's:
```python
        @wraps(func)
        def inner_decorator(*args, **kwargs):
            return func(*args, **kwargs)
```
So `inner_decorator` from line 54 in `tasks.py` "forgets" to carry the information over (ie. doesn't set `.wrapped`). From my stacktrace, `func` here points to the original function, so `_wrap_as_new` will return `inner_decorator` instead of wrapping everything in `WrappedCallableTask`. There's no way for `WrappedCallableTask` to peek inside `inner_decorator` and discover the argspec of `func`, but `@wraps` ensures that the docstring and name are right.

**Comment by kaaveland** [comment]
Clearly not the right fix as it broke the tests.

**Comment by kaaveland** [comment]
Could this be "new-style" vs "old-style" detection not working right?

**Comment by kaaveland** [comment]
Just setting `.wrapped` on `inner_decorator` seems to fix the immediate issue and not cause any tests to go red, but that smells off wrong fix to me. Anyway, think that we have found the root cause here now.

@bitprophet -- does your memory tell us anything about what the 'right' way to fix this would be?

**Comment by kaaveland** [comment]
The problem with `@parallel` looks like it's the same thing happening, but I haven't investigated that (only looked at the code).

**Comment by rafgut** [comment]
#1183 seems to be dead, #1481 is closed because it is indicating to here.
Since in version 1.13.2 this problem still occurs, I'd like to ask for more investigation into it.
